blog/2024-02-02-migrating-machines-to-xstate-v5/index.mdx

```diff
@@ -0,0 +1,565 @@
+          // Later, the action implementation function will be
+          // passed this string value as the 2nd arg, params.
+          params: ({ event }) => event.prop1,
+        },
+        {
+          type: 'doThat',
+          // Later, the action implementation function will be
+          // passed this number value as the 2nd arg, params.
+          params: ({ event }) => event.prop2,
+        },
+      ],
+    },
+  },
+  states: {
+    first: {
+      entry: {
+        type: 'whenEntering',
+        // Later, the action implementation function will be
+        // passed only these 2 values as the 2nd arg, params.
+        params: ({ event }) => ({
+          prop1: event.prop1,
+          prop2: event.prop2,
+        }),
+      },
+      exit: {
+        type: 'whenExiting',
+        // Later, the action implementation function will be
+        // passed only these 2 values as the 2nd arg, params.
+        params: ({ event }) => ({
+          prop2: event.prop2,
+          prop3: event.prop3,
+        }),
+      },
+    },
+  },
+});
+```
+
+</TabItem>
+
+<TabItem value="v4" label="XState v4">
+
+```ts
+// ❌ DEPRECATED
+import { createMachine } from 'xstate';
+
+const machine = createMachine({
+  on: {
+    next: {
+      actions: ['doThis', 'doThat'],
+    },
+  },
+  states: {
+    first: {
+      entry: 'whenEntering',
+      exit: 'whenExiting',
+    },
+  },
+});
+```
+
+</TabItem>
+</Tabs>
+
+## 4. Include implementations or stubs in `setup()`
+
+We must [provide implementations for actions, guards, and actors](/docs/machines#providing-implementations) in the object passed to `setup()`. If the machine has everything it needs to carry out these implementations on its own then these will be the actual implementations. However, if the machine needs to receive dependencies from the outside world, then these will be stubs that will be overridden later. You may end up passing a combination of concrete implementations and stubs to `setup()` which is perfectly fine.
+
+:::info
+
+When using TypeScript, the implementations passed to `setup()` must include types for the 2nd `params` argument.
+
+:::
+
+### Stubbing actions
+
+```ts
+setup({
+  actions: {
+    doThis: (_, prop1: string) => {
+      // Concrete implementation here
+      console.log(prop1);
+    },
+    // Stubbed implementation
+    doThat: (_, prop2: number) => {},
+    whenEntering: (_, params: { prop1: string; prop2: number }) => {
+      // Concrete implementation here
+      console.log(prop1, prop2);
+    },
+    // Stubbed implementation
+    whenExiting: (_, params: { prop2: number; prop3: boolean }) => {},
+  },
+});
+```
+
+:::info
+
+Stubbed actions should use an empty function body `() => {}`
+
+:::
+
+### Stubbing guards
+
+```ts
+setup({
+  guards: {
+    'is this ready': (_, prop1: string) => {
+      // Concrete implementation here
+      return prop1.length > 0;
+    },
+    // Stubbed implementation
+    'are we there yet': (_, prop2: number) => false,
```

Comment:
```suggestion
    'is this ready': (_, params: string) => {
      // Concrete implementation here
      return params.length > 0;
    },
    // Stubbed implementation
    'are we there yet': (_, params: number) => false,
```

I also recommend always using objects for params; more explicit

Reply:
I went back and forth with that. Should we recommend that users always use an object named `params` here for consistency (unless they destructure in which case there's no name)?